### PR TITLE
Add sdf -> usd light functionality to usd component

### PR DESCRIPTION
**Pull request by adlarkin**
Signed-off-by: Ashton Larkin <42042756+adlarkin@users.noreply.github.com>

# 🎉 New feature

## Summary
This builds on top of #817 to add light converting functionality from SDF to USD.

There are a few things that may need to be addressed, but they aren't blocking the core functionality (I left some TODO notes in the code to highlight things we might want to address). I think we can start to review this PR now, and then re-visit the TODOs later when time permits to add enhanced functionality. Perhaps we should create an issue that lists what enhanced functionality is missing from the USD -> SDF creator?

I also think we can close #796 because the only test in that PR is for lights, which has been moved to this PR (we will need to address https://github.com/ignitionrobotics/sdformat/pull/796#pullrequestreview-842491424, which is also a potential issue with the tests that are written in #817).

## Test it
You can run the unit test that has been added in this PR (`Light_Sdf2Usd_TEST`), or you can try converting https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo6/examples/worlds/lights.sdf to a USD file with the `sdf2usd` executable that's generated in #817 (this assumes you have PXR installed and have built the `usd` component of sdformat).

If you want to see how the light intensities look in isaac sim, you'll need to add an object to the `lights.sdf` file above in order to have the lights reflect off of something. You can add the following code to `usd/src/World.cc`, which adds a ground plane to the world:
```diff
diff --git a/usd/src/World.cc b/usd/src/World.cc
index a3aaa138..90bb81f7 100644
--- a/usd/src/World.cc
+++ b/usd/src/World.cc
@@ -30,6 +30,16 @@
 #include "sdf/World.hh"
 #include "sdf/usd/Light.hh"
 
+// includes needed for creating a plane
+#include <pxr/usd/usdGeom/xform.h>
+#include <ignition/math/Vector3.hh>
+#include <ignition/math/Pose3.hh>
+#include <pxr/usd/usdGeom/cube.h>
+#include <pxr/base/gf/vec3f.h>
+#include <pxr/base/vt/array.h>
+#include <pxr/usd/usdGeom/xformCommonAPI.h>
+#include "sdf/usd/Utils.hh"
+
 namespace usd
 {
   bool ParseSdfWorld(const sdf::World &_world, pxr::UsdStageRefPtr &_stage,
@@ -69,6 +79,23 @@ namespace usd
               << "the moment. Models that are children of the world "
               << "are currently being ignored.\n";
 
+    // temporarily adding ground plane in order to test light intensities
+    const pxr::SdfPath sdfModelPath(_path + "/plane");
+    auto usdModelXform = pxr::UsdGeomXform::Define(_stage, sdfModelPath);
+    ignition::math::Vector3d planePosition(0, 0, -0.25);
+    usd::SetPose(ignition::math::Pose3d(planePosition,
+          ignition::math::Quaterniond::Identity), _stage, sdfModelPath);
+    const pxr::SdfPath geomPath(_path + "/plane/geom");
+    auto usdCube = pxr::UsdGeomCube::Define(_stage, pxr::SdfPath(geomPath));
+    usdCube.CreateSizeAttr().Set(1.0);
+    pxr::GfVec3f endPoint(0.5);
+    pxr::VtArray<pxr::GfVec3f> extentBounds;
+    extentBounds.push_back(-1.0 * endPoint);
+    extentBounds.push_back(endPoint);
+    usdCube.CreateExtentAttr().Set(extentBounds);
+    pxr::UsdGeomXformCommonAPI cubeXformAPI(usdCube);
+    cubeXformAPI.SetScale(pxr::GfVec3f(10, 10, 0.5));
+
     return true;
   }
 }
```

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Review by ahcorde — APPROVED** [review]
LGTM, I'm not going to merge this PR yet, I'm waiting to get a review in the targeted branch

File: usd/src/Light.cc [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    if (typeParsed)
+    {
+      // TODO(adlarkin) incorporate sdf::Light's <direction> somehow? According
+      // to the USD API, things like UsdLuxDistantLight and UsdLuxDiskLight emit
+      // light along the -Z axis, so I'm not sure if this can be changed.
+      usd::SetPose(usd::PoseWrtParent(_light), _stage, sdfLightPath);
+
+      // This is a workaround to set the light's intensity attribute. Using the
+      // UsdLuxLightAPI sets the light's inputs:intensity attribute, but isaac
+      // sim reads the light's intensity attribute
```
Comment (koonpeng) [inline usd/src/Light.cc:71]: Maybe we should set both the usdLux's attribute and isaac sim's attribute in the case that we want it to work for other usd renderers.

File: usd/src/UsdTestUtils.hh [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+namespace usd
+{
+namespace testing
+{
+
+/// \brief Compare the pose of a USD prim to a desired pose
+/// \param[in] _usdPrim The USD prim
+/// \param[in] _targetPose The pose that _usdPrim should have
+void CheckPrimPose(const pxr::UsdPrim &_usdPrim,
```
Comment (koonpeng) [inline usd/src/UsdTestUtils.hh:36]: Is it possible that there will be no translate/rotate attributes if there are no relevant transforms?

**Review by koonpeng — APPROVED** [review]
LGTM, just some minor comments.

File: usd/src/UsdTestUtils.hh (hunk 2 above)
Comment (adlarkin) [reply usd/src/UsdTestUtils.hh:36]: Can you explain/give an example of what you mean by "no relevant transforms"?

File: usd/src/Light.cc (hunk 1 above)
Comment (adlarkin) [reply usd/src/Light.cc]: Good idea; I have added this in https://github.com/ignitionrobotics/sdformat/pull/818/commits/5d2c902be843059f3beb23eab08ae14c06d5b265

File: usd/src/UsdTestUtils.hh (hunk 2 above)
Comment (koonpeng) [reply usd/src/UsdTestUtils.hh:43]: I'm thinking in the case that e.g. all the translate are 0, there may not be any translate op, but now I see the code which generate the pose, the operators are always present regardless.

File: usd/src/UsdTestUtils.hh [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    EXPECT_DOUBLE_EQ(usdTranslation[0], _targetPose.Pos().X());
+    EXPECT_DOUBLE_EQ(usdTranslation[1], _targetPose.Pos().Y());
+    EXPECT_DOUBLE_EQ(usdTranslation[2], _targetPose.Pos().Z());
+    checkedTranslate = true;
+  }
+  EXPECT_TRUE(checkedTranslate);
+
+  bool checkedRotate = false;
+  if (auto rotateAttr =
+      _usdPrim.GetAttribute(pxr::TfToken("xformOp:rotateXYZ")))
```
Comment (koonpeng) [inline usd/src/UsdTestUtils.hh:61]: ignition pose is implemented with quaternions internally, but is represented by euler angles in sdf. Should we use `orient` or `rotateXYZ`?

**Comment by adlarkin** [comment]
I've updated this PR to match the USD component files that currently exist in the `sdf12` branch (from https://github.com/ignitionrobotics/sdformat/pull/817). If CI comes back green, I believe that this is ready to be merged.

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-04-13-fortress-edifice/1367/1

**Pull request merged**